WhateverGreen/kern_igfx.hpp

```diff
@@ -570,1062 +508,703 @@ class IGFX {
-
-				default:
-					return Vendor::Unknown;
-			}
+		inline LSPCON *getLSPCON(uint32_t index) {
+			return lspcons[index].lspcon;
 		}
 
 		/**
-		 *  [Vendor Helper] Get the string representation of the adapter vendor
+		 *  [Convenient] Set the non-null LSPCON driver for the given framebuffer
+		 *
+		 *  @param index A **valid** framebuffer index; Must be less than `MaxFramebufferConnectorCount`
+		 *  @param lspcon A non-null LSPCON driver instance associated with the given framebuffer
 		 */
-		static inline const char *getVendorString(Vendor vendor) {
-			switch (vendor) {
-				case Vendor::Parade:
-					return "Parade";
-
-				case Vendor::MegaChips:
-					return "MegaChips";
-
-				default:
-					return "Unknown";
-			}
+		inline void setLSPCON(uint32_t index, LSPCON *lspcon) {
+			lspcons[index].lspcon = lspcon;
 		}
 
 		/**
-		 *  [DP++ Helper] Check whether this is a HDMI adapter based on the adapter info
+		 *  [Convenient] Get the preferred LSPCON mode for the given framebuffer
 		 *
-		 *  @param info A non-null DP++ adapter info instance
-		 *  @return `true` if this is a HDMI adapter, `false` otherwise.
+		 *  @param index A **valid** framebuffer index; Must be less than `MaxFramebufferConnectorCount`
+		 *  @return The preferred adapter mode.
 		 */
-		static inline bool isHDMIAdapter(DisplayPortDualModeAdapterInfo *info) {
-			return memcmp(info->hdmiID, "DP-HDMI ADAPTOR\x04", 16) == 0;
+		inline LSPCON::Mode getLSPCONPreferredMode(uint32_t index) {
+			return lspcons[index].preferredMode;
 		}
-
+		
 		/**
-		 *  [DP++ Helper] Check whether this is a LSPCON adapter based on the adapter info
+		 *  [Wrapper] Retrieve the DPCD info for a given framebuffer port
 		 *
-		 *  @param info A non-null DP++ adapter info instance
-		 *  @return `true` if this is a LSPCON DP-HDMI adapter, `false` otherwise.
+		 *  @param that The hidden implicit `this` pointer
+		 *  @param framebuffer A framebuffer instance
+		 *  @param displayPath A display path instance
+		 *  @return `kIOReturnSuccess` on success, other values otherwise.
+		 *  @note This is a wrapper for Apple's original `AppleIntelFramebufferController::GetDPCDInfo()` method.
+		 *        Used to inject code to initialize the driver for the onboard LSPCON chip.
 		 */
-		static inline bool isLSPCONAdapter(DisplayPortDualModeAdapterInfo *info) {
-			// Guard: Check whether it is a DP to HDMI adapter
-			if (!isHDMIAdapter(info))
-				return false;
+		static IOReturn wrapGetDPCDInfo(void *that, IORegistryEntry *framebuffer, void *displayPath);
+		
+	public:
+		// MARK: Patch Submodule IMP
+		void init() override;
+		void deinit() override;
+		void processKernel(KernelPatcher &patcher, DeviceInfo *info) override;
+		void processFramebufferKext(KernelPatcher &patcher, size_t index, mach_vm_address_t address, size_t size) override;
+	} modLSPCONDriverSupport;
+	
+	/**
+	 *  The LSPCON driver requires access to I2C-over-AUX transaction APIs
+	 */
+	friend class LSPCON;
```

Comment:
Cannot private classes in C++(ones declared inside another) access private fields of their owner? Maybe I am forgetting something.

Reply:
`LSPCON` is a standalone class declared in `kern_igfx_lspcon.hpp` and is no longer nested in `IGFX`.

Reply:
Ah, fine then.